src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -467,7 +492,19 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
       // Constant eval
       //////////////////////////////////////////////////////////////////////////
       let body = '';
-      if (globalTestConfig.unrollConstEvalLoops) {
+      if (parameterTypes.some(ty => scalarTypeOf(ty).kind === 'f64')) {
+        // Directly assign the expression to the output, to avoid an
+        // intermediate store, which will concretize the value earlier
+        body = cases
+          .map(
+            (c, i) =>
+              `  outputs[${i}].value = ${toStorage(
+                resultType,
+                expressionBuilder(map(c.input, v => v.wgsl()))
+              )};`
+          )
+          .join('\n  ');
```

Comment:
Remove the `toStorage` from the `values` array, and I don't think you need this any more.